tools/uploader/s3.go

```diff
@@ -0,0 +1,318 @@
+		S3Client: client,
+	}
+
+	// 64 GB default limit
+	uploader.MaxPartUploads = util.GetEnvOrDefault(UPLOADER_MAX_PARTS_UPLOAD_KEY, 1024*4, parseInt32)
+	// 16 MB part size default
+	uploader.MultipartPartSize = util.GetEnvOrDefault(UPLOADER_MULTIPART_SIZE_KEY, 16*1024*1024, parseInt64)
+	// 3 full upload retries default
+	uploader.MaxMPURetries = util.GetEnvOrDefault(UPLOADER_MAX_MPU_RETRIES_KEY, 3, parseInt32)
+	// 30 retries per part default
+	uploader.MaxPartUploads = util.GetEnvOrDefault(UPLOADER_MAX_PART_UPLOAD_RETRIES_KEY, 30, parseInt32)
+
+	return uploader
+}
+
+func (s *s3Uploader) Initialize() {
+	if !s.initialized {
+		s.mpuBufferPool = sync.Pool{
+			New: func() interface{} {
+				b := make([]byte, s.MultipartPartSize)
+				return &b
+			},
+		}
+		s.initialized = true
+	}
+}
+
+func (s *s3Uploader) UploadObject(ctx context.Context, objectPath, bucket, key string, p ProgressConfig) error {
+	// Open the file for uploading
+	fileInfo, err := os.Stat(objectPath)
+	if err != nil {
+		return err
+	}
+
+	// Start the progress bar
+	trimDesc := p.GetPrefix()
+	bar := p.progress.New(fileInfo.Size(),
+		mpb.BarStyle().Rbound("|"),
+		mpb.PrependDecorators(
+			decor.Name(trimDesc+" "),
+			decor.Counters(decor.SizeB1024(0), "% .2f / % .2f"),
+		),
+		mpb.AppendDecorators(
+			decor.EwmaETA(decor.ET_STYLE_GO, 30),
+			decor.Name(" ] "),
+			decor.EwmaSpeed(decor.SizeB1024(0), "% .2f", 30),
+		),
+	)
+	defer bar.Abort(false)
+
+	size := fileInfo.Size()
+	fInfo := uploadFileDetails{
+		path:   objectPath,
+		size:   size,
+		bucket: bucket,
+		key:    key,
+		bar:    bar,
+	}
+	uploadMinSize := s.MultipartPartSize
+	if size <= (uploadMinSize * 3) {
+		return s.uploadObjectToS3(ctx, fInfo)
+	}
+	return s.uploadMultipartObjectToS3(ctx, fInfo)
+}
+
+func (s *s3Uploader) uploadObjectToS3(ctx context.Context, fInfo uploadFileDetails) error {
+	// Read the complete object content to the buffer so that we don't get partial data. This is a throwaway buffer but
+	// maxes out at 3x the part size since above that we use multipart upload.
+	buffer := bytes.NewBuffer(make([]byte, 0, fInfo.size))
+	file, err := os.Open(fInfo.path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	proxyReader := fInfo.bar.ProxyReader(file)
+	defer proxyReader.Close()
+	_, err = io.Copy(buffer, proxyReader)
+	if err != nil {
+		return fmt.Errorf("failed to read object content %s: %w", fInfo.path, err)
+	}
+
+	_, err = s.S3Client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket:        aws.String(fInfo.bucket),
+		Key:           aws.String(fInfo.key),
+		Body:          bytes.NewReader(buffer.Bytes()),
+		ContentLength: aws.Int64(fInfo.size),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to upload object to S3 (%s/%s): %w", fInfo.bucket, fInfo.key, err)
+	}
+
+	return nil
+}
+
+func (s *s3Uploader) uploadMultipartObjectToS3(ctx context.Context, fInfo uploadFileDetails) error {
+
+	// Initialize if we haven't already
+	if !s.initialized {
+		s.Initialize()
+	}
```

Comment:
IIUC this should be safe, right?
```suggestion
	s.Initialize()
```